t/lib/t/MusicBrainz/Server/Controller/OAuth2.pm

```diff
@@ -464,13 +482,134 @@ test 'Exchange authorization code' => sub {
         redirect_uri => 'urn:ietf:wg:oauth:2.0:oob',
         code => $code,
     });
-    $response = from_json($test->mech->content);
-    is($response->{error}, undef);
-    is($response->{error_description}, undef);
-    is($response->{token_type}, 'bearer');
-    ok($response->{access_token});
+    $response = token_response_ok($test);
     ok($response->{refresh_token});
-    ok($response->{expires_in});
+};
+
+test 'Authorize web workflow online with PKCE' => sub {
+    my $test = shift;
+
+    MusicBrainz::Server::Test->prepare_test_database($test->c, '+oauth');
+
+    my ($authorization_code, $response);
+    my $common_auth_params = '/oauth2/authorize' .
+        '?client_id=id-web' .
+        '&response_type=code' .
+        '&scope=profile' .
+        '&state=xyz' .
+        '&redirect_uri=http://www.example.com/callback';
+
+    $test->mech->get_ok('/login');
+    $test->mech->submit_form(with_fields => { username => 'editor1', password => 'pass' });
+
+    $test->mech->max_redirect(0);
+    $test->mech->get($common_auth_params . '&code_challenge_method=idk');
+    oauth_redirect_error(
+        $test->mech,
+        'www.example.com', '/callback', 'xyz',
+        'invalid_request', q(The code_challenge_method must be 'S256' or 'plain'),
+    );
+
+    $test->mech->get($common_auth_params . '&code_challenge_method=S256');
+    oauth_redirect_error(
+        $test->mech,
+        'www.example.com', '/callback', 'xyz',
+        'invalid_request', q(A code_challenge_method was supplied without an accomanying code_challenge),
+    );
+
+    $test->mech->get($common_auth_params . '&code_challenge=&code_challenge_method=S256');
+    oauth_redirect_error(
+        $test->mech,
+        'www.example.com', '/callback', 'xyz',
+        'invalid_request', q(Invalid code_challenge; see https://tools.ietf.org/html/rfc7636#section-4.1),
```

Comment:
Yup. It's unlikely the syntax highlighter supports Perl's quoting operators correctly, since you can use almost any delimiters you want - `q(foo) q[foo] q|foo| q;foo; q!foo! q^foo^ q<foo>` and more all work, and can be useful if you know none of those appear in the string. (`q` is equivalent to single quotes, for double quotes/variable substitution you'd use `qq`.)

Though I'm not sure why I didn't just use `''` on this line since the error message doesn't contain single quotes. I'll change that.